### PR TITLE
feat(openshift): Support Kind DeploymentConfig targets

### LINKED ISSUE #50: Support DeploymentConfig Kind as Target
- [x] Basic model impl
- [x] Support in e2e

### PR DESCRIPTION
**Pull request by aslakknutsen**
- [x] Double e2e test, 2 services. 1 as D and 1 as DC

fixes #50

### CODE REVIEW

File: pkg/openshift/deploymentconfig.go [hunk 1]
```diff
@@ -15,13 +22,92 @@ var _ model.Revertor = DeploymentConfigRevertor
...
+			ctx.Log.Info("Failed to delete DeploymentConfig", "name", deployment.Name)
+			ref.AddResourceStatus(model.ResourceStatus{Kind: DeploymentConfigKind, Name: deployment.Name, Action: model.ActionFailed})
+			return err
+		}
+		ref.RemoveResourceStatus(model.ResourceStatus{Kind: DeploymentConfigKind, Name: deployment.Name})
+	}
 	return nil
 }
+
+func cloneDeployment(deployment *appsv1.DeploymentConfig) *appsv1.DeploymentConfig {
```
Comment (aslakknutsen) [inline pkg/openshift/deploymentconfig.go]: This duplication should be handled in #61

File: pkg/openshift/deploymentconfig_test.go [hunk 2]
```diff
@@ -0,0 +1,208 @@
+package openshift_test
```
Comment (aslakknutsen) [inline pkg/openshift/deploymentconfig_test.go]: This is pretty much identical to k8s/deployment_test.go, same with the impl for that matter. 'maybe' worth creating a single impl that takes a 'type' handle of some sort as input? @bartoszmajsak wdyt?

File: pkg/openshift/deploymentconfig.go (hunk 1 above)
Comment (bartoszmajsak) [reply pkg/openshift/deploymentconfig.go]: Don't forget to align that with changes from #110

File: pkg/openshift/deploymentconfig.go [hunk 3]
```diff
@@ -15,13 +27,95 @@ var _ model.Revertor = DeploymentConfigRevertor
...
+	labelsClone["telepresence"] = "test"
+	deploymentClone.SetName(deployment.GetName() + "-test")
+	deploymentClone.SetLabels(labelsClone)
+	deploymentClone.Spec.Selector = labelsClone
+	deploymentClone.Spec.Template.SetLabels(labelsClone)
+	deploymentClone.SetResourceVersion("")
+	deploymentClone.Spec.Replicas = 1
+
+	container := deploymentClone.Spec.Template.Spec.Containers[0]
+	container.Image = "datawire/telepresence-k8s:0.98"
```
Comment (bartoszmajsak) [inline pkg/openshift/deploymentconfig.go]:
```suggestion
	container.Image = "datawire/telepresence-k8s:0.99"
```

File: pkg/openshift/deploymentconfig_test.go (hunk 2 above)
Comment (bartoszmajsak) [reply pkg/openshift/deploymentconfig_test.go]: From one side that would ensure we cover paths for both `Deployment` and `DeploymentConfig`. I don't know by heart if there are any substantial differences between those objects we would need to cover separately (e.g. triggers?), so not sure how much we will diverge in the implementation yet. This actually raises a question of why we don't handle both types in one set of components.

As for the implementation - DataTable tests might be a way to have parametrized input for those test cases, but I am not sure that is a fit here. We are looking for suite-level input, not really per-each functionality.

Comment (aslakknutsen) [reply pkg/openshift/deploymentconfig_test.go]: Not 100% sure how much they might diverge either, which is why I'm a little eerie to merge them. Certainly, there is an overlap on the general 'blackbox' level. 

I've got an idea, let me see if it works :)

Comment (aslakknutsen) [reply pkg/openshift/deploymentconfig_test.go]: yea... terrible idea. two sets there is :)

File: deploy/bookinfo/bookinfo.yaml [hunk 4]
```diff
@@ -28,8 +28,8 @@ spec:
   selector:
     app: details
 ---
-apiVersion: extensions/v1beta1
-kind: Deployment
+apiVersion: apps.openshift.io/v1
+kind: DeploymentConfig
```
[the remark this answers is not in the document]
Comment (aslakknutsen) [reply deploy/bookinfo/bookinfo.yaml]: True

[the remark this answers is not in the document]
Comment (aslakknutsen) [reply deploy/bookinfo/bookinfo.yaml]: so diverge, or you got a better idea?

[the remark this answers is not in the document]
Comment (aslakknutsen) [reply deploy/bookinfo/bookinfo.yaml]: I guess we could script it to in theory make the update easier.. ?

[the remark this answers is not in the document]
Comment (bartoszmajsak) [reply deploy/bookinfo/bookinfo.yaml]: I think some automated manipulation on the source would probably be the way to go.

**Comment by bartoszmajsak** [comment]
@aslakknutsen I took care of aligning new stuff with recent changes and tested manually against your TP branch. Seems to be working fine right now. I would like to streamline the process of testing `ike` <-> `telepresence` changes as right now this requires few hacks to get the images pulled and client happy. I will open an issue

**Comment by bartoszmajsak** [comment]
@aslakknutsen I took care of aligning new stuff with recent changes and tested manually against your TP branch. Seems to be working fine right now. I would like to streamline the process of testing `ike` <-> `telepresence` changes as right now this requires few hacks to get the images pulled and client happy. I will open an issue

File: deploy/bookinfo/bookinfo.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (aslakknutsen) [reply deploy/bookinfo/bookinfo.yaml]: https://github.com/Maistra/istio-workspace/pull/134

File: .golangci.yml [hunk 5]
```diff
@@ -6,7 +6,7 @@ linters-settings:
   gocyclo:
     min-complexity: 12
   dupl:
-    threshold: 128
+    threshold: 128
```
Comment (bartoszmajsak) [inline .golangci.yml]:
```suggestion
    threshold: 128
```

**Comment by aslakknutsen** [comment]
On hold awaiting Telepresence release.

File: e2e/smoke_test.go [hunk 6]
```diff
@@ -85,102 +86,121 @@ var _ = Describe("Smoke End To End Tests - against OpenShift Cluster with Istio
...
-				fmt.Printf("[%s] checking...\n", time.Now().Format("2006-01-02 15:04:05.001"))
-				return GetBodyWithHeaders("http://istio-ingressgateway-istio-system.127.0.0.1.nip.io/productpage", map[string]string{"x-test-suite": "smoke"})
-			}, 3*time.Minute, 1*time.Second).Should(ContainSubstring("Publisher Ike"))
-
-			stopFailed := ikeWithWatch.Stop()
-			Expect(stopFailed).ToNot(HaveOccurred())
-
-			Eventually(ikeWithWatch.Done(), 1*time.Minute).Should(BeClosed())
+			It("should watch for changes in ratings service in specified namespace and serve it", func() {
+				doBasicScenarioCheck(tmpDir, namespace)
```
Comment (bartoszmajsak) [inline e2e/smoke_test.go]: Maybe we can rename `doBasicScenarioCheck(tmpDir, namespace)` to `verifyThatResponseMatchesModifiedService(tmpDir, namespace)`?

File: pkg/openshift/deploymentconfig.go [hunk 7]
```diff
@@ -15,15 +29,102 @@ var _ model.Revertor = DeploymentConfigRevertor
 
 // DeploymentConfigLocator attempts to locate a DeploymentConfig kind based on Ref name
 func DeploymentConfigLocator(ctx model.SessionContext, ref *model.Ref) bool { //nolint[:hugeParam]
-	return false
+	deployment, err := getDeploymentConfig(ctx, ctx.Namespace, ref.Name)
+	if err != nil {
+		if errors.IsNotFound(err) { // Ref is not a Deployment type
```
Comment (bartoszmajsak) [inline pkg/openshift/deploymentconfig.go]:
```suggestion
		if errors.IsNotFound(err) { // Ref is not a DeploymentConfig type
```

File: pkg/openshift/openshift_suite_test.go [hunk 8]
```diff
@@ -0,0 +1,15 @@
+package openshift_test
+
+import (
+	"testing"
```
Comment (bartoszmajsak) [inline pkg/openshift/openshift_suite_test.go]:
```suggestion
	"testing"

	"go.uber.org/goleak"
```

File: pkg/openshift/openshift_suite_test.go [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+	. "github.com/maistra/istio-workspace/test"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+func TestOpenshift(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecWithJUnitReporter(t, "Openshift object Suite")
+}
```
Comment (bartoszmajsak) [inline pkg/openshift/openshift_suite_test.go]:
```suggestion
}

var _ = SynchronizedAfterSuite(func() {}, func() {
	goleak.VerifyNone(GinkgoT(),
		goleak.IgnoreTopFunction("github.com/maistra/istio-workspace/vendor/k8s.io/klog.(*loggingT).flushDaemon"),
		goleak.IgnoreTopFunction("github.com/maistra/istio-workspace/vendor/github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).registerForInterrupts"),
	)
})

```

File: e2e/smoke_test.go (hunk 6 above)
Comment (aslakknutsen) [reply e2e/smoke_test.go]: https://github.com/Maistra/istio-workspace/pull/95/commits/4ab788abfa01b34506adbdb49691823a43253f1c

**Pull request merged**